### PR TITLE
[enhancement](rebalance) Add clone task failed window to reblance tablet

### PR DESCRIPTION
**Pull request by deardeng**
## Proposed changes

Add a time window to track the failure status of the be clone task and determine whether to rebalance it

1. Clone has failed more than a certain number of times in the recent period of time
2. The TableSchedule is for replicas with missing versions, and they are on unhealthy machines. If there are enough remaining machines, and another replica can be created on other machines, the abnormal machines will be skipped when selecting dest in Tablet Sched.

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java [hunk 1]
```diff
@@ -766,7 +767,17 @@ private void handleReplicaVersionIncomplete(TabletSchedCtx tabletCtx, AgentBatch
             throws SchedException {
         stat.counterReplicaVersionMissingErr.incrementAndGet();
         try {
-            tabletCtx.chooseDestReplicaForVersionIncomplete(backendsWorkingSlots);
+            Set<Long> backendsInTablet = tabletCtx.getReplicas().stream()
+                    .map(Replica::getBackendId).collect(Collectors.toSet());
+
+            boolean skipAlwaysCloneFail = Config.create_new_replica_in_health_backends
+                    && Env.getCurrentSystemInfo().getAllBackends().stream().anyMatch(be -> be.isScheduleAvailable()
+                    && !backendsInTablet.contains(be.getId())
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:775]: add condition:   && !be.isExceedCloneFailedLimit()

**Review by yujun777 — COMMENTED** [review]
reduce cloneFailedWindow  storage

**Comment by deardeng** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/system/Backend.java [hunk 2]
```diff
@@ -823,4 +830,56 @@ public String getTagMapString() {
...
+
+        delExpireStatUnLock(windowRight);
+
+        if (cloneFailedWindow.peekLast() != null
+                && curFailTimeStamp <= cloneFailedWindow.peekLast().first + 10 * 1000) {
+            cloneFailedWindow.peekLast().second++;
+        } else {
+            cloneFailedWindow.addLast(Pair.of(curFailTimeStamp, 1L));
+        }
+        lock.unlock();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/system/Backend.java:863]: unlock should use in finally block of  'try {} finally {}';

File: fe/fe-core/src/main/java/org/apache/doris/system/Backend.java [hunk 3]
```diff
@@ -823,4 +830,56 @@ public String getTagMapString() {
...
+        lock.lock();
+        long currentTimeStamp = System.currentTimeMillis();
+        long windowRight = currentTimeStamp - Config.be_check_health_window_length * 1000L;
+
+        delExpireStatUnLock(windowRight);
+
+        long count = cloneFailedWindow.stream()
+                .filter(stat -> (stat.first <= currentTimeStamp && stat.first >= windowRight))
+                .mapToLong(stat -> stat.second).sum();
+        lock.unlock();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/system/Backend.java:876]: unlock should use in finally block of  'try {} finally {}';

File: be/src/olap/task/engine_clone_task.cpp [hunk 4]
```diff
@@ -158,6 +158,12 @@ Status EngineCloneTask::execute() {
 }
 
 Status EngineCloneTask::_do_clone() {
+    DBUG_EXECUTE_IF("EngineCloneTask.finish.random", {
+        if (rand() % 100 < (100 * dp->param("percent", 0.5))) {
+            LOG_WARNING("EngineCloneTask.finish.random random failed");
```
Comment (yujun777) [inline be/src/olap/task/engine_clone_task.cpp:163]: log add tablet id

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java [hunk 5]
```diff
@@ -1713,6 +1726,7 @@ public boolean finishCloneTask(CloneTask cloneTask, TFinishTaskRequest request)
                 }
             } else if (e.getStatus() == Status.UNRECOVERABLE) {
                 // unrecoverable
+                Env.getCurrentSystemInfo().getBackend(tabletCtx.getDestBackendId()).updateCloneFailedWindow();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1729]: delete line 1729.    unrecoverable is not really clone error,   it's cause by program bug or other error like  the clone tablet has been deleted like;

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java [hunk 6]
```diff
@@ -1725,6 +1739,7 @@ public boolean finishCloneTask(CloneTask cloneTask, TFinishTaskRequest request)
             LOG.warn("got unexpected exception when finish clone task. tablet: {}",
                     tabletCtx.getTabletId(), e);
             stat.counterTabletScheduledDiscard.incrementAndGet();
+            Env.getCurrentSystemInfo().getBackend(tabletCtx.getDestBackendId()).updateCloneFailedWindow();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1742]: delete line 1742.  reason is like line 1729

File: fe/fe-core/src/main/java/org/apache/doris/system/Backend.java [hunk 7]
```diff
@@ -823,4 +830,56 @@ public String getTagMapString() {
...
+        return count > Config.clone_tablet_in_window_failed_limit_number;
+    }
+
+    public void clearCloneFailedWindow() {
+        lock.lock();
+        cloneFailedWindow.clear();
+        lock.unlock();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/system/Backend.java:883]: unlock should use in finally block of 'try {} finally {}';

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java (hunk 5 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1729]: line 1734.  should call  clearCloneFailedWindow()  // if clone one succ, clear the whole error windowns

**Review by yujun777 — CHANGES_REQUESTED** [review]
make some changes

File: fe/fe-core/src/main/java/org/apache/doris/system/Backend.java (hunk 3 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/system/Backend.java:876]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/system/Backend.java (hunk 7 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/system/Backend.java:883]: fixed

File: be/src/olap/task/engine_clone_task.cpp (hunk 4 above)
Comment (deardeng) [reply be/src/olap/task/engine_clone_task.cpp:163]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java (hunk 5 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1729]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java (hunk 6 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1742]: fixed

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Review by yujun777 — APPROVED** [review]
LGTM

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java [hunk 8]
```diff
@@ -1777,6 +1789,7 @@ public boolean finishCloneTask(CloneTask cloneTask, TFinishTaskRequest request)
             tabletCtx.setErrMsg(e.getMessage());
             if (e.getStatus() == Status.RUNNING_FAILED) {
                 tabletCtx.increaseFailedRunningCounter();
+                Env.getCurrentSystemInfo().getBackend(tabletCtx.getDestBackendId()).updateCloneFailedWindow();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1792]: need consider getBackend = null for deleted be

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java [hunk 9]
```diff
@@ -1799,6 +1812,7 @@ public boolean finishCloneTask(CloneTask cloneTask, TFinishTaskRequest request)
             } else if (e.getStatus() == Status.FINISHED) {
                 // tablet is already healthy, just remove
                 finalizeTabletCtx(tabletCtx, TabletSchedCtx.State.CANCELLED, e.getStatus(), e.getMessage());
+                Env.getCurrentSystemInfo().getBackend(tabletCtx.getDestBackendId()).clearCloneFailedWindow();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1815]: need consider getBackend = null for deleted be

File: fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java [hunk 10]
```diff
@@ -1813,6 +1827,7 @@ public boolean finishCloneTask(CloneTask cloneTask, TFinishTaskRequest request)
         stat.counterCloneTaskSucceeded.incrementAndGet();
         gatherStatistics(tabletCtx);
         finalizeTabletCtx(tabletCtx, TabletSchedCtx.State.FINISHED, Status.FINISHED, "finished");
+        Env.getCurrentSystemInfo().getBackend(tabletCtx.getDestBackendId()).clearCloneFailedWindow();
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TabletScheduler.java:1830]: need consider getBackend = null for deleted be

**Comment by deardeng** [comment]
image
先变为draft

**Pull request closed without merging**